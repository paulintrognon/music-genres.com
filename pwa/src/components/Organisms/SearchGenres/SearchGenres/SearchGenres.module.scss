.container {
  width: 40rem;
  max-width: 100%;
  margin: auto;
  margin-top: 2rem;
}

.searchInput {
}
