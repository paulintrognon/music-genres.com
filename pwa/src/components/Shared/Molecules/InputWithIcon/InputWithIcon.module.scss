.container {
  position: relative;
  display: flex;
  align-items: center;
}

.icon {
  position: absolute;
  width: auto;
  right: 1rem;
}
