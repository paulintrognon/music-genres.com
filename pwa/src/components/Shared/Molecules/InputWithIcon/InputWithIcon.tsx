import TextInput, { Props as TextInputProps } from '../../Atoms/TextInput/TextInput'
import classnames from 'classnames'
import classes from './InputWithIcon.module.scss'

interface Props extends TextInputProps {
  icon: React.ReactNode
  onSubmit?: () => void
}
const InputWithIcon: React.FC<Props> = ({ icon, className, onSubmit, ...props }) => {
  const handleOnSubmit = (event: React.FormEvent): void => {
    event.preventDefault()
    onSubmit && onSubmit()
  }

  return (
    <form className={classnames(classes.container, className)} onSubmit={handleOnSubmit}>
      <TextInput {...props} />
      <button type="submit" className={classes.icon}>
        {icon}
      </button>
    </form>
  )
}
export default InputWithIcon
