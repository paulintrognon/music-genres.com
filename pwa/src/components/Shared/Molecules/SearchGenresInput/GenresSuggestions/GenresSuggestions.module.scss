.suggestion {
  display: block;
  width: 100%;
  padding: 0.75rem 1rem;
  background-color: var(--color-secondary);
  
  a {
    text-decoration: underline;
  }
}

.hover:hover {
  background-color: var(--color-secondary-light);
  box-shadow: 0 1px 10px 0 rgba(0,0,0,.19);
  position: relative;
}