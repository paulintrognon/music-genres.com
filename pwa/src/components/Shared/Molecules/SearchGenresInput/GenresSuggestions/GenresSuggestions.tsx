import Link from 'next/link'
import { MusicGenreBaseType } from '../../../../../types/MusicGenre/MusicGenreBaseType'
import classnames from 'classnames'
import classes from './GenresSuggestions.module.scss'
import Loader from '../../../Atoms/Loader/Loader'

type Props = {
  isLoading: boolean
  genreName: string
  suggestions: MusicGenreBaseType[]
  slugToHref: (slug: string) => string
}
const GenresSuggestions: React.FC<Props> = ({ isLoading, genreName, suggestions, slugToHref }) => {
  if (isLoading) {
    return (
      <div className={classes.suggestion} style={{ textAlign: 'center' }}>
        <Loader />
      </div>
    )
  }
  if (suggestions.length === 0) {
    return (
      <div className={classes.suggestion}>
        &quot; <i>{genreName}</i> &quot; does not exist yet.{' '}
        <Link href={`/add/video/${genreName}`}>
          <a>Create it!</a>
        </Link>
      </div>
    )
  }
  return (
    <div>
      {suggestions.map((suggestion) => (
        <Link key={suggestion.slug} href={slugToHref(suggestion.slug)}>
          <a className={classnames(classes.suggestion, classes.hover)}>{suggestion.name}</a>
        </Link>
      ))}
    </div>
  )
}
export default GenresSuggestions
