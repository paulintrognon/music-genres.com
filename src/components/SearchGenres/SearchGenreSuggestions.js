import React from 'react';
import { connect } from 'react-redux';
import { goToGenre } from '../../actions/musicGenresActions';

import './suggestions.css';
import Suggestion from './Suggestion';

function mapStoreToProps(store) {
  return {
    isFocused: store.searchGenre.isFocused,
    suggestions: store.searchGenre.suggestions,
    suggestionSelected: store.searchGenre.suggestionSelected,
  };
}
class SearchGenreSuggestions extends React.Component {
  constructor() {
    super();
    this.state = {
      text: '',
    };
    this.onClickHandler = this.onClickHandler.bind(this);
  }

  onClickHandler(value) {
    this.props.dispatch(goToGenre(value));
  }

  render() {
    if (!this.props.isFocused || !this.props.suggestions.length) {
      return '';
    }
    return (
      <ul className="suggestions-container">
        {this.props.suggestions.map((suggestion, i) => (
          <Suggestion key={i}
              value={suggestion.slug}
              name={suggestion.name}
              isSelected={i === this.props.suggestionSelected}
              onSuggestionClick={this.onClickHandler}
          ></Suggestion>
        ))}
      </ul>
    );
  }
}
export default connect(mapStoreToProps)(SearchGenreSuggestions);
