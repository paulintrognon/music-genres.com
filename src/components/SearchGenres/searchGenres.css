.search-genres-container .search-input-container {
  position: relative;
}

.search-genres-container .search-input-container .search-input {
  width: 625px;
  border: 2px solid #FFF;
  color: #FFF;
  background-color: #E62979;
  padding: 15px 25px;
  box-shadow: 0 0 5px #AAA;
  font-size: 1.5rem;
}

.search-genres-container .search-input-container .search-icon {
  position: absolute;
  right: 30px;
  bottom: 1px;
  cursor: pointer;
  background-color: #E62979;
}
