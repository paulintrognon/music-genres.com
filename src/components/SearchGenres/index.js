import React from 'react';
import { connect } from 'react-redux';
import {
  changeFocusAction,
  suggestGenresAction,
  resetGenresSuggestionsAction,
  selectSuggestion
} from '../../actions/searchGenresActions';

import SearchGenreSuggestions from './SearchGenreSuggestions';
import magnifyingGlass from './magnifying-glass.png';
import './searchGenres.css';

function mapStoreToProps(store) {
  return {
    suggestions: store.searchGenre.suggestions,
    isFocused: store.searchGenre.isFocused,
  };
}
class SearchGenre extends React.Component {
  constructor() {
    super();
    this.state = {
      text: '',
    };
  }

  onFocus() {
    this.props.dispatch(changeFocusAction(true));
  }

  onBlur() {
    if (!this.state.text.length) {
      this.props.dispatch(changeFocusAction(false));
    }
  }

  handleKeyDown(event) {
    if (event.key === 'ArrowDown') {
      this.props.dispatch(selectSuggestion('down'));
    } else if (event.key === 'ArrowUp') {
      this.props.dispatch(selectSuggestion('up'));
    }
  }

  handleChange(event) {
    const text = event.target.value;
    this.setState({ text });
    if (text.length >= 3) {
      this.props.dispatch(suggestGenresAction(text));
    } else {
      this.props.dispatch(resetGenresSuggestionsAction());
    }
  }

  render() {
    return (
      <div className="search-genres-container" onKeyDown={this.handleKeyDown.bind(this)}>
        <p>
          <span className={"search-input-container " + (this.props.isFocused ? 'focused' : '')}>
            <input
              type="text"
              className="search-input"
              placeholder="Look for a genre"
              onChange={this.handleChange.bind(this)}
              onFocus={this.onFocus.bind(this)}
              onBlur={this.onBlur.bind(this)}
              value={this.state.text}
            />
            <img src={magnifyingGlass} className="search-icon" alt="Search music genres!" />
          </span>
        </p>
        <SearchGenreSuggestions></SearchGenreSuggestions>
      </div>
    );
  }
}
export default connect(mapStoreToProps)(SearchGenre);
