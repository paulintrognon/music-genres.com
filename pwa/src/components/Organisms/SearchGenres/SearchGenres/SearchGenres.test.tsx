import React from 'react'
import createRenderer from 'react-test-renderer/shallow'
import { shallow } from 'enzyme'
import SearchGenres from './SearchGenres'
import InputWithIcon from '../../../Molecules/InputWithIcon/InputWithIcon'
import GenresSuggestions from '../GenresSuggestions/GenresSuggestions'

// Mocking useSWR
import useSWR from 'swr'
jest.mock('swr')
const useSWRMock = useSWR as jest.Mock

const slugToHref = (slug: string): string => `/path/to/${slug}`

beforeEach(() => {
  jest.resetAllMocks()
  useSWRMock.mockReturnValue({ data: { result: [] } })
})

it('should render SearchGenres', () => {
  const renderer = createRenderer.createRenderer()
  expect(renderer.render(<SearchGenres slugToHref={slugToHref} />)).toMatchSnapshot()
})

it('should search genres on input', async () => {
  const onDirty = jest.fn()
  const onClean = jest.fn()

  const component = shallow(
    <SearchGenres onDirty={onDirty} onClean={onClean} slugToHref={slugToHref} />
  )

  // On component mount, useSWR should be called with an empty string
  expect(useSWRMock).toHaveBeenCalledTimes(1)
  expect(useSWRMock.mock.calls[0]).toEqual([''])
  expect(component.find(GenresSuggestions)).toHaveLength(0)

  // On query "test" typed ...
  const suggestionsResult = [{ name: 'Genre', slug: 'genre' }]
  useSWRMock.mockReturnValueOnce({ data: { result: suggestionsResult } })
  component.find(InputWithIcon).simulate('change', {
    target: { value: 'test' },
  })
  // ...useSWR should be called with corresponding API url
  expect(useSWRMock).toHaveBeenCalledTimes(2)
  expect(useSWRMock.mock.calls[1]).toEqual(['/api/music-genres/search?query=test&limit=5'])
  // ...and suggestions should be shown
  const suggestionsComponent = component.find(GenresSuggestions)
  expect(suggestionsComponent).toHaveLength(1)
  expect(suggestionsComponent.props().suggestions).toEqual(suggestionsResult)
})

it('should call onSubmit with typed text on submit', async () => {
  const onSubmit = jest.fn()

  const component = shallow(<SearchGenres onSubmit={onSubmit} slugToHref={slugToHref} />)

  // On query "BardCore" typed ...
  component.find(InputWithIcon).simulate('change', {
    target: { value: 'BardCore' },
  })

  // And on submit...
  component.find(InputWithIcon).simulate('submit')

  // onSubmit should be called with "BardCode"
  expect(onSubmit).toHaveBeenCalledTimes(1)
  expect(onSubmit.mock.calls[0]).toEqual(['BardCore'])
})
