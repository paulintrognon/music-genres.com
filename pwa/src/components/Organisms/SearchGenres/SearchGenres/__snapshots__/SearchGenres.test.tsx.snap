// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render SearchGenres 1`] = `
<div
  className="container"
>
  <InputWithIcon
    className="searchInput"
    icon={<MagnifyingGlassIcon />}
    onChange={[Function]}
    onSubmit={[Function]}
    placeholder="Start looking for a genre"
  />
</div>
`;
